File: api/app/analysis/reporting.py
from datetime import date

from sqlalchemy.ext.asyncio import AsyncSession

from app.analysis.classifier import (
    _LEVEL_CUTS,
    WEIGHTS,
    Classifier,
    PeriodClassification,
)
from app.analysis.metrics import (
    Granularity,
    MetricsService,
    PeriodMetrics,
    bucket_start,
)
from app.schemas.reports import (
    Coverage,
    PeriodReport,
    Report,
    ReportMeta,
    WindowSummary,
)

SMALL_SAMPLE_THRESHOLD = 8


class ReportService:
    def __init__(self, session: AsyncSession) -> None:
        self._metrics = MetricsService(session)
        self._classifier = Classifier()

    async def build(
        self,
        granularity: Granularity,
        start: date | None = None,
        end: date | None = None,
    ) -> Report:
        metrics = await self._metrics.compute(granularity)
        classifications = self._classifier.classify(metrics)
        by_period: dict[date, PeriodMetrics] = {m.period: m for m in metrics}

        in_window = [
            (by_period[c.period], c)
            for c in classifications
            if _within(c.period, start, end)
        ]

        periods = [_to_report(m, c) for m, c in in_window]
        periods.reverse()

        window = _build_window(granularity, in_window, start, end)

        return Report(
            meta=ReportMeta(
                granularity=granularity.value,
                sample_size=len(metrics),
                weights=WEIGHTS,
                level_cuts={level.value: cut for cut, level in _LEVEL_CUTS},
                small_sample=len(metrics) < SMALL_SAMPLE_THRESHOLD,
                coverage=_coverage(metrics),
            ),
            window=window,
            periods=periods,
        )


def _within(period: date, start: date | None, end: date | None) -> bool:
    if start is not None and period < start:
        return False
    if end is not None and period > end:
        return False
    return True


def _coverage(metrics: list[PeriodMetrics]) -> Coverage:
    if not metrics:
        return Coverage(first=None, last=None)
    return Coverage(first=metrics[0].period, last=metrics[-1].period)


def _build_window(
    granularity: Granularity,
    in_window: list[tuple[PeriodMetrics, PeriodClassification]],
    start: date | None,
    end: date | None,
) -> WindowSummary:
    metrics = [m for m, _ in in_window]
    composite = (
        sum(c.composite for _, c in in_window) / len(in_window) if in_window else 0.0
    )
    periods = [m.period for m in metrics]
    current_bucket = bucket_start(granularity, date.today())

    return WindowSummary(
        start=start if start is not None else (min(periods) if periods else None),
        end=end if end is not None else (max(periods) if periods else None),
        level=Classifier._to_level(composite),
        composite=round(composite, 4),
        commit_count=sum(m.commit_count for m in metrics),
        pr_count=sum(m.pr_count for m in metrics),
        pr_merged=sum(m.pr_merged for m in metrics),
        additions=sum(m.additions for m in metrics),
        deletions=sum(m.deletions for m in metrics),
        active_days=sum(m.active_days for m in metrics),
        partial_current=current_bucket in periods,
    )


def _to_report(m: PeriodMetrics, c: PeriodClassification) -> PeriodReport:
    return PeriodReport(
        period=m.period,
        level=c.level,
        composite=round(c.composite, 4),
        components={k: round(v, 4) for k, v in c.components.items()},
        commit_count=m.commit_count,
        pr_count=m.pr_count,
        pr_merged=m.pr_merged,
        additions=m.additions,
        deletions=m.deletions,
        active_days=m.active_days,
        merge_rate=round(m.merge_rate, 4) if m.merge_rate is not None else None,
    )

File: packages/core/src/reporting.ts
/**
 * Report builder — TS port of `api/app/analysis/reporting.py`
 * (`ReportService`).
 *
 * Flow (identical to Python):
 *   1. classify over the FULL baseline (stable population),
 *   2. filter periods to the requested [start, end] window,
 *   3. build a WindowSummary (aggregate totals + mean-of-in-window composites
 *      → level via the same cuts; `partialCurrent` via bucketStart(today)),
 *   4. compute coverage (first/last baseline period),
 *   5. emit periods newest-first.
 */

import { bucketStart, toLocalISODate } from "./bucket.js";
import {
  classify,
  LEVEL_CUTS_BY_LEVEL,
  toLevel,
  WEIGHTS,
} from "./classifier.js";
import { computeMetrics } from "./metrics.js";
import {
  type Coverage,
  type Granularity,
  type PeriodClassification,
  type PeriodMetrics,
  type PeriodReport,
  type Report,
  mergeRate,
  type WindowSummary,
} from "./types.js";

export const SMALL_SAMPLE_THRESHOLD = 8;

/** Anything with an `.execute()` — satisfied by the Drizzle client. */
interface Executor {
  execute(query: unknown): Promise<unknown>;
}

export interface BuildReportInput {
  db: Executor;
  userId: string;
  granularity: Granularity;
  /** Inclusive window start, "YYYY-MM-DD". Omit for no lower bound. */
  start?: string;
  /** Inclusive window end, "YYYY-MM-DD". Omit for no upper bound. */
  end?: string;
}

const round4 = (value: number): number => Math.round(value * 1e4) / 1e4;

function within(period: string, start?: string, end?: string): boolean {
  if (start !== undefined && period < start) {
    return false;
  }
  if (end !== undefined && period > end) {
    return false;
  }
  return true;
}

function coverageOf(metrics: PeriodMetrics[]): Coverage {
  if (metrics.length === 0) {
    return { first: null, last: null };
  }
  return {
    first: metrics[0]!.period,
    last: metrics[metrics.length - 1]!.period,
  };
}

function toPeriodReport(
  m: PeriodMetrics,
  c: PeriodClassification,
): PeriodReport {
  const rate = mergeRate(m);
  return {
    period: m.period,
    level: c.level,
    composite: round4(c.composite),
    components: {
      throughput: round4(c.components.throughput),
      active_days: round4(c.components.active_days),
      churn: round4(c.components.churn),
    },
    commitCount: m.commitCount,
    prCount: m.prCount,
    prMerged: m.prMerged,
    additions: m.additions,
    deletions: m.deletions,
    activeDays: m.activeDays,
    mergeRate: rate === null ? null : round4(rate),
  };
}

/**
 * Pure window summary — aggregates the in-window periods and derives the
 * representative level. Factored out of `buildReport` so it can be unit-tested
 * with hand-built data and no DB. `today` is injected for determinism.
 *
 * Mirrors `_build_window` in the Python reporting module exactly.
 */
export function summarizeWindow(
  granularity: Granularity,
  inWindow: Array<{ metrics: PeriodMetrics; classification: PeriodClassification }>,
  start: string | undefined,
  end: string | undefined,
  today: Date,
): WindowSummary {
  const metrics = inWindow.map((x) => x.metrics);
  const composite =
    inWindow.length > 0
      ? inWindow.reduce((acc, x) => acc + x.classification.composite, 0) /
        inWindow.length
      : 0;
  const periods = metrics.map((m) => m.period);
  const currentBucket = bucketStart(granularity, toLocalISODate(today));

  const sum = (pick: (m: PeriodMetrics) => number): number =>
    metrics.reduce((acc, m) => acc + pick(m), 0);

  const minPeriod = periods.length > 0 ? periods.reduce((a, b) => (a < b ? a : b)) : null;
  const maxPeriod = periods.length > 0 ? periods.reduce((a, b) => (a > b ? a : b)) : null;

  return {
    start: start !== undefined ? start : minPeriod,
    end: end !== undefined ? end : maxPeriod,
    level: toLevel(composite),
    composite: round4(composite),
    commitCount: sum((m) => m.commitCount),
    prCount: sum((m) => m.prCount),
    prMerged: sum((m) => m.prMerged),
    additions: sum((m) => m.additions),
    deletions: sum((m) => m.deletions),
    activeDays: sum((m) => m.activeDays),
    partialCurrent: periods.includes(currentBucket),
  };
}

/**
 * Builds the full report for `userId`. The only DB-touching call is
 * `computeMetrics`; all derivation below it is pure (and shared with
 * `summarizeWindow`). `today` is injectable for tests.
 */
export async function buildReport(
  input: BuildReportInput,
  today: Date = new Date(),
): Promise<Report> {
  const { db, userId, granularity, start, end } = input;

  const metrics = await computeMetrics(db, userId, granularity);
  const classifications = classify(metrics);
  const byPeriod = new Map(metrics.map((m) => [m.period, m]));

  const inWindow = classifications
    .filter((c) => within(c.period, start, end))
    .map((c) => ({ metrics: byPeriod.get(c.period)!, classification: c }));

  const periods = inWindow
    .map((x) => toPeriodReport(x.metrics, x.classification))
    .reverse();

  const window = summarizeWindow(granularity, inWindow, start, end, today);

  return {
    meta: {
      granularity,
      sampleSize: metrics.length,
      weights: { ...WEIGHTS },
      levelCuts: { ...LEVEL_CUTS_BY_LEVEL },
      smallSample: metrics.length < SMALL_SAMPLE_THRESHOLD,
      coverage: coverageOf(metrics),
    },
    window,
    periods,
  };
}
